designs/2021-suppression-support/README.md

```diff
@@ -0,0 +1,452 @@
+    "fixableWarningCount":0,
+    "usedDeprecatedRules":[]
+  }
+]
+```
+
+### Add new CLI option `--track-suppressions`
+
+We propose to add a new CLI option `--track-suppressions` for the new feature. Without the option, ESLint will work as it is currently, and `suppressions` property should not exist in the output. With this option, ESLint is expected to export all violations, including suppressed ones. `suppressions` should be _empty_ for non-suppressed violations, while `suppressions` should be a list for suppressed ones, according to [suppressions property](https://docs.oasis-open.org/sarif/sarif/v2.0/csprd02/sarif-v2.0-csprd02.html#_Toc10127852).
+
+### Parse description in comments as `Justification`
+
+We propose not to ignore the text after two or more dashes in comments and considered it as the justification of a suppression. The modification should be in `getDirectiveComments`.
+
+### Add suppressions attribute in `LintMessage`
+
+`LintMessage` is the type that finally output to the formatter. Currently it contains `ruleId`, `severity`, `message`, etc. What we desire is a `suppressions` attribute in `LintMessage` when the CLI option `--track-suppressions` used. It should be a list that contains a suppression entry (`{kind: "directive", justification: "Fake justification message."}`).
+
+### Reserve directive suppression
+
+In the step of apply directive suppression, ESLint will not remove the problems that should be disabled by directive comments if it gets the `--track-suppressions` option. These problems will be reserved and added `{kind: "directive", justification: "Justification message example."}`.
+
+NOTE that the kind `directive` should be converted to `inSource` for SARIF according to [kind property](https://docs.oasis-open.org/sarif/sarif/v2.1.0/os/sarif-v2.1.0-os.html#_Toc34317735).
+
+### Modify `CLIEngine` and related formatters
+
+As mentioned, we are not going to change the current behavior of ESLint. Without `--no-inline-config`, ESLint should not report errors/warnings in CMD whether with `--track-suppressions` option or not. In `CLIEngine`, when `suppressions` of a message is not null or empty (which means this message should be suppressed and not be reported), this message would be skipped or ignored. Additionally, the related formatters, such as stylish, should also be modified.
+
+## Documentation
+
+Add examples and explanations of the option `--track-suppressions` to _User guide – Command Line Interface_.
+
+## Drawbacks
+
+- Additional contexts must be transferred across modules, such as `justification` in `directives` and `suppressions` in `LintMessage`.
+- Related formatters, such as stylish, should be modified to keep the current behavior.
+
+## Backwards Compatibility Analysis
+
+Basically, current users would not feel any difference, since the proposed feature would only work with the new option. Without the option `--track-suppressions`, any current behaviors would not be changed.
```

Comment:
Depending on how third-party formatters process `LintMessage`. Only using the new option `--track-suppressions`, there could be incompatible problems. Without the option, nothing will be affected.

Reply:
If user runs eslint with `--track-suppressions` and an existing third-party formatter, it will show suppressed errors as if they were not suppressed. I think this is a compatibility issue we should address.

Reply:
I'm not sure what would be the best approach.

Some thoughts:

* Instead of mixing suppressed and non-suppressed problems in [messages](https://eslint.org/docs/developer-guide/working-with-custom-formatters#the-results-object), maybe we could add a new property `suppressedMessages` (array). In that case, an option might even be unnecessary? (we could just always provide them and it would be up to the formatter to merge and show all, or entirely ignore `suppressedMessages` as all the existing formatters are doing now).
* Adding an ability for formatters to define features they support, and throw an error if the given formatter doesn't support suppressions. If the formatter doesn't support supressions, that wouldn't mean that it is incompatible and that it needs to release a new version, just that it isn't designed to be used with `--track-suppressions`.

Reply:
If we just always return all suppressed problems in a separate `suppressedMessages` property on [LintResult](https://eslint.org/docs/developer-guide/nodejs-api#-lintresult-type), that would avoid situations where user runs `eslint --track-suppressions` and sees nothing due to a formatter that doesn't support this feature, because there would be no `--track-suppressions` option.

Reply:
If I'm a JSON formatter user, I would feel strange that so many suppressed messages come to my output after updating ESLint... Because I suppressed so I don't want to see those violations.

Maybe an explicit option could give users a clear idea whether tracking suppressions is working.

Reply:
I think it’s okay to always return the suppressed messages. The JSON formatter is primarily used by automation to transform the results output into another format so it’s unlikely to cause problems.

Reply:
So seems like we have an agreement here, which is we will not need to add the --track-suppressions switch and always output the suppressions to a separate property (suppressedMessages). Formatters that know about this new property can merge and handle the messages accordingly, and for formatters that don't then it is an NoOp.

Agreed?

If we all agree, we will move forward to making the code changes and submit PR for reviews when we are ready.

Thanks.

Reply:
Hi @mdjermanovic @nzakas @DSIE63 , the RFC is updated and the option related things are removed now.

Reply:
@DSIE63 please wait until the RFC has been approved before submitting code.